clients/xload/xload.man
.TH XLOAD 1 "25 October 1988" "X Version 11"
.SH NAME
xload - load average display for X
.SH SYNOPSIS
.B xload
[-\fItoolkitoption\fP ...] [-scale \fIinteger\fP] [-update \fIseconds\fP]
.SH DESCRIPTION
The 
.I xload 
program displays a periodically updating histogram of the system load average.
This program is nothing more than a wrapper around the Athena Load widget.
.SH OPTIONS
.PP
.I Xload
accepts all of the standard X Toolkit command line options along with the 
following additional options:
.PP
.TP 8
.B \-scale \fIinteger\fP
This option specifies the minimum number of tick marks in the histogram,
where one division represents one load average point.  If the load goes
above this number, \fIxload\fP will create more divisions, but it will never
use fewer than this number.  The default is 1.
.PP
.TP 8
.B \-update \fIseconds\fP
This option specifies the frequency in seconds at which
.I xload
updates its display.  If the load average window is uncovered (by moving 
windows with a window manager or by the \fIxrefresh\fP program), the graph
will be also be updated.  The minimum amount of time allowed between updates
is 5 seconds (which is also the default).
.TP 8
.B \-hl \fIcolor\fP
This option specifies the color of the label and scale lines.  
.PP
The following standard X Toolkit arguments are commonly used
with \fIxload\fP:
.PP
.TP 8
.B \-bd \fIcolor\fP
This option specifies the border color.
The default is \fIblack\fP.
.PP
.TP 8
.B \-bg \fIcolor\fP
This option specifies the background color.
The default is \fIwhite\fP.
.PP
.TP 8
.B \-bw \fIpixels\fP
This option specifies the width in pixels of the border around the window.
The default value is 2.
.PP
.TP 8
.B \-fg \fIcolor\fP
This option specifies the graph color.
The default color is \fIblack\fP.
.PP
.TP 8
.B \-fn \fIfontname\fP
This option specifies the font to be used in displaying the name of the 
host whose load is being monitored.  The default is \fI6x10\fP.
.PP
.TP 8
.B \-rv
This option indicates that reverse video should be simulated by swapping the
foreground and background colors.
.PP
.TP 8
.B \-geometry \fIgeometry\fP
This option specifies the prefered size and postion of the window.
.PP
.TP 8
.B \-display \fIdisplay\fP
This option specifies the X server to contact.
.PP
.TP 8
.B \-xrm \fIresourcestring\fP
This option specifies a resource string to be used.  This is especially
useful for setting resources that do not have separate command line options.
.SH "X DEFAULTS"
.PP
This program uses the 
.I Load
widget in the X Toolkit.  It understands all of the core resource names and
classes as well as:
.TP 8
.B width (\fPclass\fB Width)
Specifies the width of the load average graph.
.TP 8
.B height (\fPclass\fB Height)
Specifies the height of the load average graph.
.TP 8
.B update (\fPclass\fB Interval)
Specifies the frequency in seconds at which the load should be redisplayed.
.TP 8
.B scale (\fPclass\fB Scale)
Specifies the initial number of ticks on the graph.  The default is 1.
.TP 8
.B minScale (\fPclass\fB Scale)
Specifies the minimum number of ticks that will be displayed.  The default
is 1.
.TP 8
.B foreground (\fPclass\fB Foreground)
Specifies the color for the graph.  
The default is \fIblack\fP since the core default for background 
is \fIwhite\fP.
.TP 8
.B highlight (\fPclass\fB Foreground)
Specifies the color for the text and scale lines.  The default is the same
as for the \fBforeground\fP resource.
.TP 8
.B label (\fPclass\fB Label)
Specifies the label to use on the graph.  The default is the hostname.
.TP 8
.B font (\fPclass\fB Font)
Specifies the font to be used for the label.  The default is \fIfixed\fP.
.TP 8
.B reverseVideo (\fPclass\fB ReverseVideo)
Specifies that the foreground and background should be reversed.
.SH SEE ALSO
X(1), xrdb(1), mem(4), Athena Load widget
.SH DIAGNOSTICS
Unable to open display or create window. Unable to open /dev/kmem.
Unable to query window for dimensions. Various X errors.
.SH BUGS
This program requires the ability to open and read the special system
file \fI/dev/kmem\fP.  Sites that do not allow general access to this file
should make \fIxload\fP belong to the same group as \fI/dev/kmem\fP and
turn on the \fIset group id\fP permission flag.
.PP
Reading /dev/kmem is inherently non-portable.  Therefore, the widget upon
which this application is based must be ported to each new operating system.
.PP
Border color has to be explicitly specified when reverse video is used.
.SH COPYRIGHT
Copyright 1988, Massachusetts Institute of Technology.
.br
See \fIX(1)\fP for a full statement of rights and permissions.
.SH AUTHORS
K. Shane Hartman (MIT-LCS) and Stuart A. Malone (MIT-LCS);
.br
with features added by Jim Gettys (MIT-Athena), Bob Scheifler (MIT-LCS),
and Tony Della Fera (MIT-Athena)


CHANGES
The following changes have been made to this directory since R2:

     o  substantially improved the X man page.
     o  added README file with quick instructions for those in the fast lane.
     o  added RELEASENOTES file giving overview of what has changed and where
	to look for more information.
     o  added STANDARDS file describing what is and is not part of the X
	Standard.
     o  removed support for automatically building contrib directory from
	the Imakefile since it won't work completely on any single platform.
     o  moved rgb directory here from util.
     o  added some misc. utility targets to the top-level Makefile.


clients/xload/xload.c
#ifndef lint
static char rcsid[] = "$XConsortium: xload.c,v 1.16 88/10/18 14:07:18 swick Exp $";
#endif  lint

#include <X11/Intrinsic.h>
#include <X11/Xatom.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include <X11/Load.h>
#include <stdio.h> 
#include "xload.bit"

extern void exit();

/* Command line options table.  Only resources are entered here...there is a
   pass over the remaining options after XtParseCommand is let loose. */

static XrmOptionDescRec options[] = {
{"-scale",	"*load.minScale",	XrmoptionSepArg,	 NULL},
{"-update",	"*load.update",		XrmoptionSepArg,	 NULL},
{"-hl",		"*load.highlight",	XrmoptionSepArg,	 NULL},
{"-highlight",	"*load.highlight",	XrmoptionSepArg,	 NULL},
};


/* Exit with message describing command line format */

void usage()
{
    fprintf(stderr,
"usage: xload [-fn {font}] [-update {seconds}] [-scale {integer}] [-rv]\n"
);
    fprintf(stderr,
"             [-geometry [{width}][x{height}][{+-}{xoff}[{+-}{yoff}]]] [-display [{host}]:[{vs}]]\n"
);
    fprintf(stderr,
"             [-fg {color}] [-bg {color}] [-bd {color}] [-bw {pixels}]\n");
    exit(1);
}

#ifndef lint
/* this silliness causes the linker to include the VendorShell
 * module from Xaw, rather than the one from Xt.
 */
static Junk()
{
#include <X11/Vendor.h>
WidgetClass junk = vendorShellWidgetClass;
}
#endif

void main(argc, argv)
    int argc;
    char **argv;
{
    char host[256];
    Widget toplevel;
    Widget load;
    Arg arg;
    char *labelname = NULL;
    Pixmap icon_pixmap = None;
    
    toplevel = XtInitialize("main", "XLoad", options, XtNumber(options), &argc, argv);
      
    if (argc != 1) usage();
    
    XtSetArg (arg, XtNiconPixmap, &icon_pixmap);
    XtGetValues(toplevel, &arg, 1);
    if (icon_pixmap == None) {
	XtSetArg(arg, XtNiconPixmap, 
		 XCreateBitmapFromData(XtDisplay(toplevel),
				       XtScreen(toplevel)->root,
				       xload_bits, xload_width, xload_height));
	XtSetValues (toplevel, &arg, 1);
    }

    XtSetArg (arg, XtNlabel, &labelname);
    load = XtCreateManagedWidget ("load", loadWidgetClass, toplevel, NULL, 0);
    XtGetValues(load, &arg, 1);
    if (!labelname) {
       (void) gethostname (host, 255);
       XtSetArg (arg, XtNlabel, host);
       XtSetValues (load, &arg, 1);
    }
    XtRealizeWidget (toplevel);
    XtMainLoop();
}
